test/e2e/syncer/spec_test.go

```diff
@@ -0,0 +1,254 @@
+/*
+Copyright 2021 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	coreexternalversions "k8s.io/client-go/informers"
+	kubernetesclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kcp-dev/kcp/test/e2e/framework"
+	"github.com/kcp-dev/kcp/test/e2e/reconciler/cluster/apis/wildwest"
+	wildwestv1alpha1 "github.com/kcp-dev/kcp/test/e2e/reconciler/cluster/apis/wildwest/v1alpha1"
+	wildwestclientset "github.com/kcp-dev/kcp/test/e2e/reconciler/cluster/client/clientset/versioned"
+	wildwestexternalversions "github.com/kcp-dev/kcp/test/e2e/reconciler/cluster/client/informers/externalversions"
+)
+
+func init() {
+	utilruntime.Must(wildwestv1alpha1.AddToScheme(scheme.Scheme))
+}
+
+//go:embed *.yaml
+var rawCustomResourceDefinitions embed.FS
+
+const crdName = "cowboys.wildwest.dev"
+const clusterName = "us-east1"
+const sourceClusterName, sinkClusterName = "source", "sink"
+
+func TestSpecSyncer(t *testing.T) {
+	type runningServer struct {
+		framework.RunningServer
+		wildwestClientSet wildwestclientset.Interface
+		k8sClientSet      kubernetesclientset.Interface
+	}
+	var testCases = []struct {
+		name string
+		work func(ctx context.Context, t framework.TestingTInterface, servers map[string]runningServer)
+	}{
+		{
+			name: "create a resource in new namespace, expect ns name to contain lcluster + ns name and for resource to get synced to correct ns",
+			work: func(ctx context.Context, t framework.TestingTInterface, servers map[string]runningServer) {
+				const testNamespace = "syncer-test"
+				expectedSinkNsName := fmt.Sprintf("kcp--admin--%s", testNamespace)
+
+				sourceNsClient := servers[sourceClusterName].k8sClientSet.CoreV1().Namespaces()
+				sourceCowboyClient := servers[sourceClusterName].wildwestClientSet.WildwestV1alpha1().Cowboys(testNamespace)
+				sinkNsClient := servers[sinkClusterName].k8sClientSet.CoreV1().Namespaces()
```

Comment:
(No action required) What is the motivation for the source/sink var naming? vs something like source/dest or <other>? sink makes sense to me in the context of event producer/consumer but resource syncing between pcluster and kcp is necessarily bidirectional (spec one way, status another. maybe metadata both ways).

cc: @stevekuznetsov

Reply:
I think this was how the syncer code labeled these concepts? I created the original test as a port of a shell script